doc/api/async_hooks.md

```diff
@@ -0,0 +1,671 @@
+# Async Hooks
+
+
+> Stability: 2 - Stable
+
+
+The `async-hooks` module provides an API to register callbacks tracking the
+lifetime of asynchronous resources created inside a Node.js application.
+
+At its simplest form it can track metadata about each resource and log it when
+the program exits.
+
+## Terminology
+
+An async resource represents either a "handle" or a "request".
+
+Handles are a reference to a system resource. Some resources are a simple
+identifier. For example, file system handles are represented by a file
+descriptor. Other handles are represented by libuv as a platform abstracted
+struct, e.g. `uv_tcp_t`. Each handle can be continually reused to access and
+operate on the referenced resource.
+
+Requests are short lived data structures created to accomplish one task. The
+callback for a request should always and only ever fire one time. Which is when
+the assigned task has either completed or encountered an error. Requests are
+used by handles to perform these tasks. Such as accepting a new connection or
+writing data to disk.
+
+## Public API
+
+### Overview
+
+Here is a simple overview of the public API. All of this API is explained in
+more detail further down.
+
+```js
+const async_hooks = require('async_hooks');
+
+// Return the id of the current execution context. Useful for tracking state
+// and retrieving the handle of the current trigger without needing to use an
```

Comment:
> ... retrieving the handle of the current trigger

That comment dates back to when I had implemented a map to keep track of all resources, and there was `async_hooks.getResourceById()`. Which would return the resource from the map, but it ended up being too costly. So it was removed.